### PR TITLE
allow multiple query parameters with same name

### LINKED ISSUE #90: multiple query parameters with same name
If you have an query string with several parameters having the same name, e.g.:

    /test?name=x&name=y

you would expect the following parameter object:

    { name: ['x', 'y'] }

but all values except the first will be ignored and you get:

    { name: 'x' }

Minimal example:

    crossroads = require('crossroads'); 

    r = crossroads.addRoute("/test:?query:", function(params){
      console.log(params);    
    });

    crossroads.parse("/test?name=x&name=y");

### PR DESCRIPTION
**Pull request by sjhewitt**
This commit was written by Kamaruni, I'm just submitting it as a pull request to get it merged into master

fixes #90

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
     "description" : "Flexible router which can be used in multiple environments",
     "keywords" : ["routes", "event", "observer", "routing", "router"],
     "homepage" : "http://millermedeiros.github.com/crossroads.js/",
-    "version" : "0.12.0",
+    "version" : "0.13.0",
```
Comment (millermedeiros) [inline package.json]: version bump should be handled during the package release (we run `npm version` before each deploy)

**Comment by millermedeiros** [comment]
main problem that I see is that I think there is no standard for multiple parameters with same name.. I've seen `?foo[]=bar&foo[]=baz` (PHP does it this way by default) and `?foo=bar&foo=baz` and `?foo=bar,baz`

**Comment by sjhewitt** [comment]
I don't see how it's a problem... This solves the problem of getting multivalued parameters from the querystring, not how to choose a way the developer should name their parameters.
It is a PHP quirk that it requires you to use names that end in `[]` to specify an array. In the comma separated case, it is the developer that has chosen to encode their values as comma separated, and they should split it themselves.

The simplest case that i'm trying to solve is:
given a form like this, that submits using get:
```
<form method="">
<input type="hidden" name="foo" value="bar">
<input type="hidden" name="foo" value="baz">
<input type="submit">
</form>
```
Use crossroads to parse the resulting querystring


btw, I've removed the version commit.

File: dev/src/intro.js [hunk 2]
```diff
@@ -83,11 +83,20 @@
...
             n = queryArr.length,
             obj = {},
-            item, val;
+            item, val, decodedVal;
         while (n--) {
             item = queryArr[n].split('=');
             val = shouldTypecast ? typecastValue(item[1]) : item[1];
-            obj[item[0]] = (typeof val === 'string')? decodeURIComponent(val) : val;
+            decodedVal = (typeof val === 'string')? decodeURIComponent(val) : val;
+            if (isArray(obj[item[0]])) {
```
Comment (millermedeiros) [inline dev/src/intro.js]: what about storing the `obj[item[0]]` on a variable:

```js
propertyName = item[0];
prev = obj[propertyName];
if (isArray(prev)) {
  prev.unshift(decodedVal);
}
else if (prev) {
  obj[propertyName] = [decodedVal, prev];
}
else {
  obj[propertyName] = decodedVal;
}
```

**Comment by millermedeiros** [comment]
@sjhewitt I agree with you, this is good as is. no need to worry about other weird use cases.

**Comment by sjhewitt** [comment]
Updated with the suggested changes

**Comment by millermedeiros** [comment]
thanks!

**Comment by quickstep25** [comment]
I've been using this pull request version for a while now.  It was much needed and is most appreciated.  So far I've had no issues in using it, works perfectly for the app I am using it with.

**Comment by hendrikstier** [comment]
One question regarding one of the weird use cases with "[]" denoting the array-type of the parameter (here: used with Rails): how can I prevent the parameter name from being URL-encoded?

For example: `...&param%255B%255D=1&... vs ...&param[]=1&`

**Pull request merged**